✨ Use device serial if everything else fails

muchzill4 opened the pull request:
### Description

If product code or online id can't be found in HTM files, resort to first 4 characters of the serial number. This should fix the K66F.

### Test Coverage

<!--
Please put an `x` in the correct box e.g. `[x]` to indicate the testing coverage of this change.
-->

- [x]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

rwalton-arm commented on tests/_internal/test_resolve_target.py:
```diff
@@ -78,15 +70,19 @@ def test_will_resolve_targets_using_online_id_when_available(
-    def test_raises_when_no_information_found_on_candidate(
-        self, read_product_code, read_online_id, _get_all_htm_files_contents
+    @mock.patch("mbed_devices._internal.resolve_target.get_target_by_product_code")
+    def test_will_resolve_targets_by_product_code_from_serial_number_if_htm_file_has_no_information(
```
The name of this test is a bit too long for me, how about "test_falls_back_to_serial_number_resolver" and put any additional information in a docstring.

rwalton-arm commented on mbed_devices/_internal/resolve_target.py:
```diff
@@ -50,7 +43,10 @@ def _build_target_resolver(candidate: CandidateDevice) -> Callable:
         return functools.partial(get_target_by_online_id, slug=online_id.device_slug, target_type=online_id.device_type)
     except OnlineIdNotFound:
-        raise UnableToBuildResolver
+        pass
```
I'd like an explicit comment as to why we `pass` here; for future me when I forget what this is doing.

rwalton-arm commented on mbed_devices/_internal/resolve_target.py:
```diff
@@ -50,7 +43,10 @@ def _build_target_resolver(candidate: CandidateDevice) -> Callable:
+        pass
+
+    # Product code might be the first 4 characters of the serial number
+    return functools.partial(get_target_by_product_code, candidate.serial_number[:4])
```
I think this `functools.partial` approach is too complex and I should've flagged it last review. Why not just have resolve target call the right function instead of this `build_target_resolver` fancy stuff? Maybe I'm missing something...

muchzill4 replied:
Oh, you're right. That'll actually work better. 👍

rosiemiller2 commented on mbed_devices/_internal/resolve_target.py:
```diff
@@ -50,7 +43,10 @@ def _build_target_resolver(candidate: CandidateDevice) -> Callable:
-        raise UnableToBuildResolver
+        pass
+
+    # Product code might be the first 4 characters of the serial number
```
```suggestion
    # Check if the first 4 characters of the serial number is the product code. It is only in some cases (eg. ST boards).
```
(I am not sure if ST boards are, just for example).

rosiemiller2 replied:
Hmm, actually maybe this is worse. Its' not really a check. I retract my comment!

rwalton-arm commented on mbed_devices/_internal/resolve_target.py:
```diff
@@ -17,60 +16,59 @@ class NoTargetForCandidate(Exception):
+    """
+    all_files_contents = _get_all_htm_files_contents(candidate.mount_points)
+
+    # 1/
```
Seems like overkill to have these comments here. I can read the code, which is fairly clear in this case.

muchzill4 replied:
https://github.com/ARMmbed/mbed-devices/pull/51/commits/f115ce593c8a51029d9b957fe9a32970df3b371e

rwalton-arm commented on mbed_devices/_internal/resolve_target.py:
```diff
@@ -17,60 +16,59 @@ class NoTargetForCandidate(Exception):
 
-def _extract_product_code(all_files_contents: Iterable[str]) -> str:
+def _extract_product_code(all_files_contents: Iterable[str]) -> Optional[str]:
+    """Return first product code fuond in files contents, None if not found."""
```
```suggestion
    """Return first product code found in files contents, None if not found."""
```

rwalton-arm commented on tests/_internal/test_resolve_target.py:
```diff
@@ -9,81 +9,101 @@
     _get_all_htm_files_contents,
     resolve_target,
-    _build_target_resolver,
+    # _build_target_resolver,
```
Should this be removed?

muchzill4 replied:
https://github.com/ARMmbed/mbed-devices/pull/51/commits/baf517933386467f62605f01e19a2d2aa37d6151

muchzill4 commented:
@rwalton-arm I've addressed your comments - did a bit of refactor based on your suggestions. I think it reads better now. Codeclimate isn't happy though, so do tell if you think we can ignore this one.

- removed `functools.partial` - now we call `mbed_targets.get_*` instead of returning it as pre-populated callable
- removed some exceptions - they added no value and made the code unnecessarily lengthy/nested
- separated giga test into `TestCases` based on each scenario

rwalton-arm commented on tests/_internal/test_resolve_target.py:
```diff
@@ -9,81 +9,101 @@
+@mock.patch("mbed_devices._internal.resolve_target.read_product_code", autospec=True, return_value=None)
+@mock.patch("mbed_devices._internal.resolve_target.read_online_id", autospec=True, return_value=None)
+@mock.patch("mbed_devices._internal.resolve_target.get_target_by_product_code", autospec=True)
+class TestResolveTargetUsingProductCodeFromSerial(TestCase):
```
I'm not convinced we need a test class per resolution case. How about `TestResolveTargets` with a test case for each case.

muchzill4 replied:
I prefer it as is tbh. Each case has certain mock state that's valid for entire case, so it's easier to group them that way.

muchzill4 commented on tests/_internal/test_resolve_target.py:
```diff
@@ -6,87 +6,103 @@
     _get_all_htm_files_contents,
     resolve_target,
-    _build_target_resolver,
+    # _build_target_resolver,
```
```suggestion
```
